### PR TITLE
Ensure compatibility with Python 2

### PR DESCRIPTION
**Pull request by brianyu28**

### CODE REVIEW

File: submit50 [hunk 1]
```diff
@@ -22,15 +21,25 @@ import tempfile
 import textwrap
 import time
 import traceback
-import urllib.request
 
+from backports.shutil_get_terminal_size import get_terminal_size
```
Comment (dmalan) [inline submit50]: I don't see a change to `setup.py` in this PR?

File: submit50 [hunk 2]
```diff
@@ -22,15 +21,25 @@ import tempfile
...
 timestamp = ""
 
+# ensure compatibility with Python 2
+try:
+   input = raw_input
+except NameError:
+   pass
+if not hasattr(shlex, "quote"):
+    shlex.quote = pipes.quote
+
```
Comment (dmalan) [inline submit50]: Would it be conventional to put this inside of the `if __name__...` block? Just curious.

File: submit50 [hunk 3]
```diff
@@ -110,8 +119,8 @@ def authenticate():
         cprint("GitHub password:", end=" ", flush=True)
         password = str()
         while True:
-            ch = getch.getch()
-            if ch == "\n": # Enter
+            ch = getch()
+            if ch in ["\n", "\r"]: # Enter
```
Comment (dmalan) [inline submit50]: If user inputs `\r\n` (Windows), do we need to consume the trailing `\n`?

File: submit50 [hunk 4]
```diff
@@ -136,11 +145,11 @@ def authenticate():
...
-        raise Error("Invalid username and/or password.") from None
+        raise Error("Invalid username and/or password.")
 
     # check for other error
     elif res.status_code != 200:
-        raise Error("Could not authenticate user.") from None
+        raise Error("Could not authenticate user.")
 
     # canonicalize (capitalization of) username, especially if user logged in via email address
     username = res.json()["login"]
```
Comment (dmalan) [inline submit50]: Should this be in a `try` in case this key doesn't exist for some reason?

File: submit50 [hunk 5]
```diff
@@ -174,7 +183,7 @@ def checkout(args):
     try:
         username, password, email = authenticate()
     except:
-        raise Error("Invalid username and/or password.") from None
+        raise Error("Invalid username and/or password.")
```
Comment (dmalan) [inline submit50]: I think you'll want to do:

```
e = Error("Invalid username and/or password.") 
e.__cause__ = None
raise e
```

File: submit50 [hunk 6]
```diff
@@ -239,7 +248,9 @@ def checkout(args):
 
 def cprint(text="", color=None, on_color=None, attrs=None, **kwargs):
     """Colorizes text (and wraps to terminal's width)."""
-    columns, lines = shutil.get_terminal_size()
+    columns, lines = get_terminal_size()
+    if sys.version_info[0] < 3 and "flush" in kwargs:
```
Comment (dmalan) [inline submit50]: Let's add a comment explaining.

File: submit50 [hunk 7]
```diff
@@ -343,7 +354,7 @@ def submit(problem):
         if run.verbose:
             cprint(str(e))
 
-        raise Error("Invalid problem. Did you mean to submit something else?") from None
+        raise Error("Invalid problem. Did you mean to submit something else?")
```
Comment (dmalan) [inline submit50]: Same `__cause__` comment (and anywhere else you re-raise an exception inside an `except:`).

File: submit50 [hunk 8]
```diff
@@ -467,11 +478,47 @@ def two_factor(username, password):
...
+class _Getch:
+    """Gets a single character from standard input."""
+    def __init__(self):
+        try:
+            self.impl = _GetchWindows()
+        except ImportError:
+            self.impl = _GetchUnix()
+
+    def __call__(self): return self.impl()
+
```
Comment (dmalan) [inline submit50]: Maybe remove these blank lines since all three classes are from same citation?

**Comment by kzidane** [comment]
Just rebased to `origin/develop`. Hope I didn't break anything, @brianyu28? Tests look good now. Let me know if we should add more!

**Comment by dmalan** [comment]
@brianyu28 should we merge this or are you waiting for more changes?

**Comment by brianyu28** [comment]
@dmalan This should be ready to merge!

**Pull request merged**